### ISSUE TITLE
Demand fails to account for negative numbers

### ISSUE THREAD

**Issue by jschwerdlin** [jcalbert/uEagle#1]
The code for interpreting the demand number assumes it will be an unsigned number.  However, with Solar, the overall demand can be negative.  So, the number needs to be interpreted as 2's complement.  This library forms the basis for the Eagle component in Home Assistant, so this percolates up to the display there.

**Comment by jcalbert** [comment]
Nice catch, I'll have a look soon.

**Comment by jschwerdlin** [comment]
Thanks.  I created an ugly hack in a custom component in Home Assistant for now.  I convert back into packed binary as an unsigned number, then back to a signed number.  I think the critical piece of code is this:
demand=unpack('>i', bytes_demand) which will unpack the binary as 2's complement. Then you can apply the multiplier and divisor

And of course:
from struct import unpack

**Comment by jcalbert** [comment]
@jschwerdlin can you tell me how you generate `bytes_demand`?  The API provides demand as a 3-byte signed int, which `unpack` doesn't seem able to handle.

**Comment by jschwerdlin** [comment]
I only had access to the Rainforest component in Home Assistant (it imports your code), so I had to hack it there.  I basically reversed the process that I figured had been done in your code. Basically I repack the value treating it as an unsigned integer, and then unpack it again as signed.  This is the full set of code

```
        if resp["Demand"] > 1000.0:
            """Value is too large, it must be negative"""
            """Convert number back to an integer, asume multiplier 1 and divisor 1000"""
            int_demand = int(resp["Demand"] * 1000)
            """pack integer"""
            bytes_demand=pack('>I',int_demand)
            """unpack as a signed integer"""
            new_int_demand=unpack('>i', bytes_demand)
            """convert baack to float, again assuming multiplier of 1 and divisor of 1000"""
            new_float_demand=new_int_demand[0]/1000.0
            out["instantanous_demand"] = new_float_demand
```

**Comment by jschwerdlin** [comment]
I think you need to work on this code section

```
    if 'Demand' in d:
        d['Demand'] = int(d['Demand']) * factor
```

**Comment by jcalbert** [comment]
Ok, I think I have a fix in the branch [twos-comp-fix](https://github.com/jcalbert/uEagle/tree/twos-comp-fix), but I don't have solar panels with which to test on actual hardware. Would you be able to try it out once the sun comes up?

**Comment by jschwerdlin** [comment]
I looked at your code and ran some offline tests with your update and it seems like it will accomplish what is needed (can't say I've ever seen that method to handle 2's complement).

Unfortunately, I'm not great with Python and even worse with GitHub (even though I was a SW engineer). The code in Home Assistant imports yours with this line:

from uEagle import Eagle as LegacyReader

I'm not sure how to have it pull in the branch instead

Also, I probably won't be home while there's still sun in the sky for a couple of days, so it's going to be a bit before I can test.

**Comment by jcalbert** [comment]
It's probably easiest to test outside of home assistant.  One way would be:

1. Clone the repository or, since all the code is in a single file, you could just [download it from here](https://raw.githubusercontent.com/jcalbert/uEagle/twos-comp-fix/uEagle/uEagle.py) and drop it into any old directory.

2. Minimal code to test the function would be:

```python3
from uEagle import Eagle
 
CLOUD_ID = '012abc' #Your cloud ID
INSTALL_CODE = '0123456789abcdef' #Your install code
ADDRESS = '192.168.1.123' # your device's address
 
eagle = Eagle(CLOUD_ID, INSTALL_CODE, address=ADDRESS)
 
demand_data = eagle.get_instantaneous_demand()['InstantaneousDemand']
print('Current Demand: {:.3f} kW'.format(demand_data['Demand']))
```

or, if you want test while you're away:

```python3
import time
from uEagle import Eagle

TIME_BETWEEN_CALLS = 60 * 10 # 10 minutes

CLOUD_ID = '012abc' #Your cloud ID
INSTALL_CODE = '0123456789abcdef' #Your install code
ADDRESS = '192.168.1.123' # your device's address

eagle = Eagle(CLOUD_ID, INSTALL_CODE, address=ADDRESS)

print('Time       Demand')
while True:
    demand_data = eagle.get_instantaneous_demand()['InstantaneousDemand']
    timestring = time.strftime('%H:%M:%S')
    print('{}   {:.3f} kW'.format(time.strftime('%H:%M:%S'), demand_data['Demand']))
    time.sleep(TIME_BETWEEN_CALLS)
```

**Comment by jschwerdlin** [comment]
OK - I've got it set up standalone (thanks for the sample code).  I'll let it run thru tomorrow and we'll see what it looks like

**Comment by jschwerdlin** [comment]
Ugg, the last comment didn't post when I thought it would.  It unfortunately doesn't seem to have worked:

08:05:12   0.347 kW
08:15:13   0.242 kW
08:25:13   0.114 kW
08:35:14   4278189.777 kW
08:45:14   4278189.559 kW
08:55:15   4278189.368 kW
09:05:16   4278189.239 kW
09:15:16   4278189.575 kW
09:25:17   4278188.806 kW
09:35:17   4278188.569 kW
09:45:18   4278188.264 kW
09:55:19   4278187.916 kW
10:05:19   4278188.649 kW

**Comment by jschwerdlin** [comment]
Trying a few code mods here.  I'll let it run thru tomorrow and see what I get

**Comment by jcalbert** [comment]
Well this is certainly... disheartening

**Comment by jcalbert** [comment]
Tomorrow I'll write up a few tests to check that the EAGLE isn't sending bad data, and try to reproduce the bug with artificial data.

**Comment by jschwerdlin** [comment]
I think this fixes it.  I basically had this code running today and saw negative values (although I had missed the [0] index on the result of the unpack.  Obviously you don't need the prints :)  Two's complement (pack/unpack) won't affect the positive numbers

What I can't tell is whether the Summation values might also suffer from this

def convert_demand(d):
    'Parse values and remove extraneous keys from demand responses.'
    factor = max(int(d['Multiplier'], 0), 1) / max(int(d['Divisor'], 0), 1)

    if 'Demand' in d:
        demand = int(d['Demand'],0)
        print(demand, ' demand before adjustment')
        bytes_demand=pack('>I',demand)
        new_int_demand=unpack('>i', bytes_demand)[0]
        print(new_int_demand, ' demand after my adjustment')
        d['Demand'] = demand * factor

**Comment by jschwerdlin** [comment]
Well, I did my first pull request with a commit on a fork of your twos complement branch.  I confirmed that code works today.  Before I pulled out the print statements, I saw this:

4294964027  demand before adjustment
-3269  demand after my adjustment
13:06:13   -3.269 kW
4294963925  demand before adjustment
-3371  demand after my adjustment
13:16:14   -3.371 kW
4294964120  demand before adjustment
-3176  demand after my adjustment
13:26:15   -3.176 kW
4294964008  demand before adjustment
-3288  demand after my adjustment
13:36:15   -3.288 kW
246  demand before adjustment
246  demand after my adjustment
13:46:16   0.246 kW
4294967171  demand before adjustment
-125  demand after my adjustment
13:56:17   -0.125 kW

**Comment by jcalbert** [comment]
Ok, the plot thickens:  `4294964027` is `0xfffff33b`. The Legacy Eagle outputs a 3 byte int but that's a (normal) 4 byte int.  As a test, could you please run this against your hardware?

```python3
eagle.get_network_info()['NetworkInfo']
```

**N.B. For privacy, remove `DeviceMacId`, `InstallCode`, `LinkKeyHigh`, and `LinkKeyLow` if they appear in the output.**

**Comment by jschwerdlin** [comment]
{'DeviceMacId': '0xXXXXXX', 'InstallCode': '0xXXXXXX', 'LinkKeyHigh': '0xXXXXXX', 'LinkKeyLow': '0xXXXXX', 'FWVersion': '1.4.48 (6952)', 'HWVersion': '1.2.6', 'Manufacturer': 'Rainforest Automation, Inc.', 'ModelId': 'Z109-EAGLE', 'DateCode': '2015040124620816', 'ImageType': '0x1301', 'Protocol': 'Zigbee'}

**Comment by jcalbert** [comment]
OK, well this is a bit of a mystery.  I'm going to create a function that handles arbitrary data sizes, but separately I should figure out why our devices give different outputs.

**Comment by jschwerdlin** [comment]
I added a print out of the raw hex (i.e. d['Demand']) and I indeed get a 6 hex digit number when the number is positive (like you apparently)

0x000373
883  demand before adjustment
883  demand after my adjustment
23:33:01   0.883 kW

The pack/unpack method seems to work fine for that length as well.  Tomorrow I'll be able to see the negative value in hex once the sun is up

**Comment by jschwerdlin** [comment]
And with the sun up

0xfffff564
4294964580  demand before adjustment
-2716  demand after my adjustment
10:28:41   -2.716 kW

**Comment by jcalbert** [comment]
This is very helpful, thank you.  Looks like they truncate the two leading hex chars if they're zero (and the developer docs are wrong 😕).  Let me make a last check or two and then I'll bring in the PR.  Thanks for all your help on this!

**Comment by jcalbert** [comment]
Resolved with PR #4
